evalme/metrics.py

```diff
@@ -0,0 +1,244 @@
+            if s > 1 or s < 0:
+                logger.error(f'Error in project {project}. Matching score {s} is not within [0, 1] interval '
+                             f'for the following results:\nx={result_first}\ny={result_second}. It will be clipped')
+                return 0 if s < 0 else 1
+            return s
+
+        if per_label:
+            for l in score:
+                if n[l] > 0:
+                    score[l] /= float(n[l])
+                else:
+                    score[l] = 0
+                score[l] = clipped(score[l])
+            return score
+
+        return clipped(score / float(n) if n > 0 else 0)
+
+    @classmethod
+    def average(cls, project, results):
+        n = len(results)
+        if n == 0:
+            return 0.0
+        if n == 1:
+            return 1.0
+        mean_score = 0
+        for i in range(n):
+            for j in range(i + 1, n):
+                mean_score += Metrics.apply(project, results[i], results[j], symmetric=True)
+        norm = n * (n - 1) / 2
+        return mean_score / max(norm, 1)
+
+    @classmethod
+    def group(cls, project, results, threshold=0.0, use_single_linkage_by_default=False):
+        """
+        Returns dict with groups, where key are group index and values are list of results indices
+        :param results: list of results
+        :param threshold: the comparison between 2 results should be above this threshold to join them in one group
+        :return: {0: [1,2,3], 1: [4,5], 2: [6]}
+        """
+        num_results = len(results)
+        if num_results == 0:
+            raise ValueError(f'Can\'t group empty results for project {project}')
+        if num_results == 1:
+            return {0: [0]}
+        result_sim = np.zeros((num_results, num_results), dtype=np.float64)
+        for i in range(num_results):
+            for j in range(i + 1, num_results):
+                result_sim[i, j] = Metrics.apply(project, results[i], results[j], symmetric=True)
+        result_sim = result_sim + result_sim.T
+        margin = 1.01 * np.max(result_sim)
+        dists = margin - squareform(result_sim)
+        if project.agreement_method == project.SINGLE or not project.agreement_method:
+            linkage_matrix = single(dists)
+        elif project.agreement_method == project.COMPLETE:
+            linkage_matrix = complete(dists)
+        else:
+            if use_single_linkage_by_default:
+                linkage_matrix = single(dists)
+            else:
+                raise ValueError(f'Unknown agreement method {project.agreement_method}')
+        clusters = fcluster(linkage_matrix, t=margin - threshold, criterion='distance')
+        groups = defaultdict(list)
+        for i, cluster_idx in enumerate(clusters):
+            groups[cluster_idx].append(i)
+        return groups
+
+
+Metrics.register(
+    name='iou_bboxes',
+    form=None,
+    tag='RectangleLabels',
+    func=iou_bboxes,
+    desc='IOU for bounding boxes'
+)
```

Comment:
Can we also add mAP as a bbox metric to use?

Reply:
Matcher.get_mAP_score()